pip_audit/_dependency_source/resolvelib/resolvelib.py

```diff
@@ -92,3 +105,48 @@ class ResolveLibResolverError(DependencyResolverError):
     """
 
     pass
+
+
+def _build_dependee_map(candidates: list[Candidate]) -> dict[_Requirement, list[_Requirement]]:
+    """
+    Build a mapping of dependee `Requirement`s to dependers. This is needed to find the top-level
+    requirements that each subdependency originates from.
+    """
+    dependee_map = {}
+    for c in candidates:
+        for dep in c.dependencies:
+            if dep not in dependee_map:
+                dependee_map[dep] = []
+            dependee_map[dep].extend(c.reqs)
+    return dependee_map
+
+
+def _find_origin_reqs(
```

Comment:
@woodruffw How does this approach look to you? I store the requirements for each candidate. Then I make a mapping to go from each candidate's dependencies back to those requirements. That way, I can work backwards from each candidate until I get to a top-level requirement.

Reply:
Looks reasonable to me!

My only request would be that we give it a slightly more descriptive name 🙂 -- it's a reverse-dependency lookup on the first level of the dependency tree, so maybe `_dependees_for_candidate` or similar?

Reply:
[5eeef40](https://github.com/pypa/pip-audit/pull/488/commits/5eeef40aff20073c1721bfe9712d864eae5e4342)